include/dca/math/nfft/dnfft_1d_gpu.hpp

```diff
@@ -0,0 +1,245 @@
+// Copyright (C) 2018 ETH Zurich
+// Copyright (C) 2018 UT-Battelle, LLC
+// All rights reserved.
+//
+// See LICENSE.txt for terms of usage.
+// See CITATION.txt for citation guidelines if you use this code for scientific publications.
+//
+// Author: Giovanni Balduzzi (gbalduzz@gitp.phys.ethz.ch)
+//
+// This class implements the 1D delayed-NFFT (d-NFFT) algorithm on the GPU.
+// See dnfft_1d.hpp for references and CPU implementation.
+
+#ifndef DCA_HAVE_CUDA
+#error "This file requires CUDA."
+#endif
+
+#ifndef DCA_MATH_NFFT_DNFFT_1D_GPU_HPP
+#define DCA_MATH_NFFT_DNFFT_1D_GPU_HPP
+
+#include "dca/math/nfft/dnfft_1d.hpp"
+
+#include <cuda_runtime.h>
+
+#include "dca/linalg/matrix.hpp"
+#include "dca/linalg/util/allocators.hpp"
+#include "dca/linalg/util/cuda_event.hpp"
+#include "dca/linalg/vector.hpp"
+#include "dca/math/nfft/kernels_interface.hpp"
+#include "dca/phys/domains/quantum/electron_band_domain.hpp"
+
+namespace dca {
+namespace math {
+namespace nfft {
+// dca::math::nfft::
+
+template <typename ScalarType, typename WDmn, typename RDmn, int oversampling = 8, NfftModeNames mode = CUBIC>
+class Dnfft1DGpu;
+
+// Only the CUBIC mode is implemented.
+template <typename ScalarType, typename WDmn, typename RDmn, int oversampling>
+class Dnfft1DGpu<ScalarType, WDmn, RDmn, oversampling, CUBIC>
+    : public Dnfft1D<ScalarType, WDmn,
+                     func::dmn_variadic<func::dmn_0<phys::domains::electron_band_domain>,
+                                        func::dmn_0<phys::domains::electron_band_domain>, RDmn>,
+                     oversampling, CUBIC> {
+public:
+  using BDmn = func::dmn_0<phys::domains::electron_band_domain>;
+  using PDmn = func::dmn_variadic<BDmn, BDmn, RDmn>;
+
+  using ThisType = Dnfft1DGpu<ScalarType, WDmn, RDmn, oversampling>;
+  using ElementType = ScalarType;
+  using BaseClass = Dnfft1D<ScalarType, WDmn, PDmn, oversampling, CUBIC>;
+
+public:
+  Dnfft1DGpu(double beta, cudaStream_t stream, bool accumulate_m_sqr = false);
+
+  // Resets the accumulated quantities. To be called before each DCA iteration.
+  void initialize();
+
+  // Accumulates asynchronously on the device the entries of the M matrix at times and orbitals
+  // described by the provided configuration.
+  // Postcondition: M and config shall not be modified untill 'synchronizeCopy' is called.
```

Comment:
typo: untill --> until

Reply:
Fixed in 547604f176339765f2a44aac02708f7b0ec5dcb7